File: Day05/C++/stack_using_queue.cpp
#include <iostream>
#include <queue>
using namespace std;

class StackUsingQueue {
    queue<int> main;
    queue<int> temp;

public:
    void push(int val) {
        while (!main.empty()) {
            temp.push(main.front());
            main.pop();
        }
        main.push(val);
        while (!temp.empty()) {
            main.push(temp.front());
            temp.pop();
        }
    }

    int pop() {
        int val = main.front();
        main.pop();
        return val;
    }

    bool isEmpty() {
        return main.empty();
    }
};

int main() {
    StackUsingQueue s;
    s.push(10);
    s.push(20);
    s.push(30);
    s.push(40);
    s.push(50);
    while (!s.isEmpty()) {
        cout << "Popped: " << s.pop() << endl; // 50, 40, 30, 20, 10
    }
    return 0;
}

File: Day05/Python/stackusingqueue.py
from collections import deque

class StackUsingQueue:
    def __init__(self):
        self.main = deque()
        self.temp = deque()

    def push(self, val):
        while self.main:
            self.temp.append(self.main.popleft())
        self.main.append(val)
        while self.temp:
            self.main.append(self.temp.popleft())

    def pop(self):
        return self.main.popleft()

    def isEmpty(self):
        return len(self.main) == 0

if __name__ == "__main__":
    s = StackUsingQueue()
    s.push(10)
    s.push(20)
    s.push(30)
    s.push(40)
    s.push(50)
    while not s.isEmpty():
        print("Popped:", s.pop())  # 50, 40, 30, 20, 10
